pyaedt/edb_core/edb_data/sources.py

```diff
@@ -506,14 +534,308 @@ def pec_launch_width(self):
+        """Get the closest pin padstack instances and serves any edge terminal connected to a pad.
+
+        Parameters
+        ----------
+        gnd_net_name_preference : str, optional
+            Preferred reference net name. Optianal, default is `None` which will auto compute the gnd name.
+
+        Returns
+        -------
+        :class:`pyaedt.edb_core.edb_data.padstacks_data.EDBPadstackInstance`
+        """
+        comp_inst = self._edb_terminal.GetComponent()
+        pins = self._pedb.core_components.get_pin_from_component(comp_inst.GetName())
+        try:
+            edges = self._edb_terminal.GetEdges()
+        except AttributeError:
+            return None
+        _, pad_edge_pstack_inst, pad_edge_layer, pad_edge_polygon_data = edges[0].GetParameters()
+        return self._get_closest_pin(pad_edge_pstack_inst, pins, gnd_net_name_preference)
+
+    @pyaedt_function_handler()
+    def _get_closest_pin(self, ref_pin, pin_list, gnd_net=None):
+        _, pad_stack_inst_point, rotation = ref_pin.GetPositionAndRotation()  # get the xy of the padstack
+        if gnd_net is not None:
+            power_ground_net_names = [gnd_net]
+        else:
+            power_ground_net_names = [net for net in self._pedb.core_nets.power_nets.keys()]
+        comp_ref_pins = [i for i in pin_list if i.GetNet().GetName() in power_ground_net_names]
+        if len(comp_ref_pins) == 0:
+            self._pedb.logger.error(
+                "Terminal with PadStack Instance Name {} component has no reference pins.".format(ref_pin.GetName())
+            )  # pragma: no cover
+            return None  # pragma: no cover
+        closest_pin_distance = None
+        pin_obj = None
+        for pin in comp_ref_pins:  # find the distance to all the pins to the terminal pin
+            if pin.GetName() == ref_pin.GetName():  # skip the reference psi
+                continue  # pragma: no cover
+            _, pin_point, rotation = pin.GetPositionAndRotation()
+            distance = pad_stack_inst_point.Distance(pin_point)
+            if closest_pin_distance is None:
+                closest_pin_distance = distance
+                pin_obj = pin
+            elif closest_pin_distance < distance:
+                continue
+            else:
+                closest_pin_distance = distance
+                pin_obj = pin
+        if pin_obj:
+            return EDBPadstackInstance(pin_obj, self._pedb)
+
+
+class ExcitationSources(CommonExcitation):
+    """Manages sources properties.
+
+    Parameters
+    ----------
+    pedb : pyaedt.edb.Edb
+        Edb object from Edblib.
+    edb_terminal : Ansys.Ansoft.Edb.Cell.Terminal.EdgeTerminal
+        Edge terminal instance from edblib.
+
+
+    Examples
+    --------
+
+    """
+
+    def __init__(self, pedb, edb_terminal):
+        CommonExcitation.__init__(self, pedb, edb_terminal)
+
+    @property
+    def magnitude(self):
```

Comment:
Documentation is missing for this property.

Reply:
thanks, I've added it to all the missing points